rfd/0028-cluster-config-resources.md

```diff
@@ -0,0 +1,298 @@
+subsequent Teleport versions for marshalling purposes.
+
+Documentation should be updated accordingly.
+
+### Restricting to a subset of values of a field
+
+Some users should not be capable of setting a field to any of the otherwise
+accepted values.  It is proposed to satisfy this requirement by introducing new
+RBAC verbs, interpreted as implying additional constraints. Such new verbs
+would be evaluated together with other RBAC verbs in the auth API handlers.
+
+In particular, this is needed for `SessionRecordingConfig` which has two
+fields:
+* `mode` with the possible values: `off`, `node`, `node-sync`,
+  `proxy`, `proxy-sync`;
+* a boolean `proxy_checks_host_keys`.
+
+Some users should be subject to the following restrictions:
+1. The value of `mode` can be changed only if its original value
+   is one of `off`, `node` or `node-sync`.
+1. The value of `mode` can be changed to only one of `off`, `node`
+   or `node-sync`.
+1. The value of `proxy_checks_host_keys` cannot be changed.
+
+This would be achieved by a new RBAC verb `update.no_proxy_recording`. A user could then
+be assigned a role featuring `update.no_proxy_recording` instead of the regular, more
+permissive `update`:
+```
+  allow:
+    rules:
+    - resources: [session_recording_config]
+      verbs: [read, update.no_proxy_recording]
+```
+
+### Additional dynamically configurable resources
+
+In addition to the resources derived from `ClusterConfig`, the resources
+`ClusterAuthPreference` and `PAMConfig` should also be adapted to facilitate
+the dynamic configuration workflow.
+
+### Working with a whole cluster configuration
+
+`KindClusterConfig` should be retained but reinterpreted as a helper meta-kind
+similar to `KindConnectors`. It would allow aggregating all the cluster config
+related resources into a single `tctl get` output:
+
+```
+$ tctl get cluster_config
+kind: session_recording_config
+[...]
+---
+kind: cluster_networking_config
+[...]
+---
+kind: audit_config
+[...]
+---
+kind: cluster_auth_preference
+[...]
+---
+kind: pam_config
+[...]
+```
+
+This combined output can be subsequently edited and used to replace the old
+configuration by passing it to the `tctl create` command which is able to
+consume multiple resource definitions (provided the user has the privileges
+needed to update all the resource kinds).
+
+Note that if a field of a configuration resource is omitted from the YAML, the
+field's value will be reset to its default.  The `tctl` workflow supports only
+replacing (or overwriting) of a stored resource with another full resource,
+not a partial update of the stored resource.
+
+### Examples
+
+#### Setting up node-sync session recording
+
+In this example, `session_recording_config` is assumed to be already
+dynamically pre-configured: in particular, while the `mode` field would be by
+default set to `node`, here it is already set to `off`.
+
+```
+$ tctl get session_recording_config | tee reccfg.yaml
```

Comment:
```
$ tctl get cluster_config
$ tctl get cluster_config session_recording
$ tctl get cluster_config pam
$ tctl get cluster_config auth_preference
```

Looks nicer

Reply:
I don't think `cluster_config` is necessary here

Reply:
That was if we treat `cluster_config` as a parent element but if each is a separate resource then yes it's not necessary.

Btw, I would also consider using a path convention to name these resources: 

```
kind: cluster_config
version: v2
spec:

kind: cluster_config/pam
version: v2
spec:

kind: cluster_config/auth_preferences
version: v2
spec:
```

Having this hierarchy potentially can make RBAC simpler too.  

```
- resources:
      - cluster_config/*
      verbs:
      - list
      - create
      - read
      - update
      - delete
```      

I understand that it might not be possible, sharing it just as an idea.